oxidation/libparsec/crates/platform-async/examples/eyeballs.rs

```diff
@@ -0,0 +1,137 @@
+        Mode::Fail(Duration::from_millis(300)),
+        Mode::Success(Duration::from_millis(300)),
+        Mode::Success(Duration::from_millis(100)),
+    ];
+
+    log::debug!("starting with {} targets", targets.len());
+    let start_time = Instant::now();
+    assert_eq!(Ok(2), dbg!(eyeballs(targets).await));
+    let elapsed = start_time.elapsed();
+    log::info!("elapsed time {}ms", elapsed.as_millis());
+    assert!(elapsed < Duration::from_millis(900));
+    Ok(())
+}
+
+async fn eyeballs(targets: Vec<Mode>) -> Result<usize, ()> {
+    let failed_attempt = targets
+        .iter()
+        .map(|_| Arc::new(Notify::new()))
+        .collect::<Vec<_>>();
+    let targets = Arc::new(targets);
+    let join_set = Vec::with_capacity(targets.len());
+    let join_set = Arc::new(Mutex::new(join_set));
+    let (tx, mut rx) = mpsc::channel::<usize>(targets.len());
+
+    spawn_attempt(0, targets, failed_attempt, join_set.clone(), tx);
+    let result = dbg!(rx.recv().await);
+    for task in join_set.lock().unwrap().iter() {
+        task.cancel();
+    }
+    result.ok_or(())
+}
+
+fn spawn_attempt(
+    which: usize,
+    targets: Arc<Vec<Mode>>,
+    failed_attempt: Vec<Arc<Notify>>,
+    join_set: Arc<Mutex<Vec<Task<()>>>>,
+    sender: mpsc::Sender<usize>,
+) {
+    log::debug!("spawning task #{which}");
+
+    let task = spawn(attempt(
+        which,
+        targets,
+        failed_attempt,
+        join_set.clone(),
+        sender,
+    ));
+
+    join_set.lock().unwrap().push(task);
+}
+
+async fn attempt(
+    which: usize,
+    targets: Arc<Vec<Mode>>,
+    failed_attempt: Vec<Arc<Notify>>,
+    join_set: Arc<Mutex<Vec<Task<()>>>>,
+    sender: mpsc::Sender<usize>,
+) {
+    log::info!("[#{which}] starting task");
+    if which > 0 {
+        let attempt_to_wait = which - 1;
+        tokio::select! {
+            _ = tokio::time::sleep(Duration::from_millis(TIMEOUT)) => {
+                log::info!("[#{which}] previous task timed out");
+            }
+            _ = failed_attempt[attempt_to_wait].notified() => {
+                log::info!("[#{which}] previous task finished");
+            }
+        }
+    }
+
+    if which + 1 < targets.len() {
+        let next_attempt = which + 1;
+        log::info!("[#{which}] start next task");
+        spawn_attempt(
+            next_attempt,
+            targets.clone(),
+            failed_attempt.clone(),
+            join_set.clone(),
+            sender.clone(),
+        );
+    }
+
+    let load = targets[which];
+    log::info!("[#{which}] executing load {load:?}");
+    match mock_load(load).await {
+        Err(_) => {
+            log::info!("[#{which}] finished with error");
+            failed_attempt[which].notify_waiters();
+        }
+        Ok(_) => {
+            log::info!("[#{which}] finished with success");
+            sender.send(which).await.unwrap();
+        }
+    }
+}
+
+async fn mock_load(mode: Mode) -> Result<(), ()> {
```

Comment:
Better to use an Enum as Result to describe more clearly

Reply:
`Result` is an `Enum`, what would you prefer ?

Reply:
```Rust
MockResult {
    Success,
    Failed,
}
```
Result is an Enum, but generic Enum, why should I want to think about if there is something wrapped inside when I read a match arms ?
It's like why should I use Enum instead of bool ^^

Reply:
You can also keep your Mode Enum
Do you know this pattern ?
```Rust
enum Enum {
    Foo(i32),
    Bar(i32),
}

fn fun() -> Enum {
    let e = Enum::Foo(5);
    let (Enum::Foo(delay) | Enum::Bar(delay)) = e;
    // sleep delay
    e
}
```

Reply:
I'm wondering what's the best way for this type of structure
```rust
struct Foo {
    delay: i32
    mode: Mode
}
enum Mode {
    Success,
    Fail,
```
or the current implementation

Reply:
```rust
let e = Enum::Foo(5);
let (Enum::Foo(delay) | Enum::Bar(delay)) = e;
```
I know this pattern

Reply:
> I'm wondering what's the best way for this type of structure
> 
> ```rust
> struct Foo {
>     delay: i32,
>     mode: Mode,
> }
> enum Mode {
>     Success,
>     Fail,
> }
> ```
> 
> or the current implementation

I like this structure, but keep in mind that this is an example